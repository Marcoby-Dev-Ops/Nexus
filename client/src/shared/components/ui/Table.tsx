import React from 'react';
import { cn } from '@/shared/lib/utils';

export interface Column {
  Header: string;
  accessor: string;
}

export interface TableProps {
  columns: Column[];
  data: Record<string, React.ReactNode>[];
  className?: string;
}

/**
 * Simple table component for 1.0
 */
const TableComponent: React.FC<TableProps> = ({ columns, data, className }) => {
  return (
    <div className={cn('overflow-x-auto', className)}>
      <table className="min-w-full divide-y divide-border">
        <thead className="bg-muted">
          <tr>
            {columns.map((column, index) => (
              <th 
                key={index}
                className="px-6 py-3 text-left text-xs font-medium text-foreground uppercase tracking-wider"
              >
                {column.Header}
              </th>
            ))}
          </tr>
        </thead>
        <tbody className="bg-background divide-y divide-border">
          {data.map((row, rowIndex) => (
            <tr key={rowIndex} className="hover: bg-muted/50">
              {columns.map((column, colIndex) => (
                <td 
                  key={colIndex}
                  className="px-6 py-4 whitespace-nowrap text-sm text-foreground"
                >
                  {row[column.accessor]}
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

// Named exports for shadcn/ui style usage
export const Table = React.forwardRef<
  HTMLTableElement,
  React.HTMLAttributes<HTMLTableElement>
>(({ className, ...props }, ref) => (
  <div className="relative w-full overflow-auto">
    <table
      ref={ref}
      className={cn("w-full caption-bottom text-sm", className)}
      {...props}
    />
  </div>
));
Table.displayName = "Table";

export const TableHeader = React.forwardRef<
  HTMLTableSectionElement,
  React.HTMLAttributes<HTMLTableSectionElement>
>(({ className, ...props }, ref) => (
  <thead ref={ref} className={cn("[&_tr]:border-b", className)} {...props} />
));
TableHeader.displayName = "TableHeader";

export const TableBody = React.forwardRef<
  HTMLTableSectionElement,
  React.HTMLAttributes<HTMLTableSectionElement>
>(({ className, ...props }, ref) => (
  <tbody
    ref={ref}
    className={cn("[&_tr: last-child]:border-0", className)}
    {...props}
  />
));
TableBody.displayName = "TableBody";

export const TableRow = React.forwardRef<
  HTMLTableRowElement,
  React.HTMLAttributes<HTMLTableRowElement>
>(({ className, ...props }, ref) => (
  <tr
    ref={ref}
    className={cn(
      "border-b transition-colors hover: bg-muted/50 data-[state=selected]:bg-muted",
      className
    )}
    {...props}
  />
));
TableRow.displayName = "TableRow";

export const TableHead = React.forwardRef<
  HTMLTableCellElement,
  React.ThHTMLAttributes<HTMLTableCellElement>
>(({ className, ...props }, ref) => (
  <th
    ref={ref}
    className={cn(
      "h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0",
      className
    )}
    {...props}
  />
));
TableHead.displayName = "TableHead";

export const TableCell = React.forwardRef<
  HTMLTableCellElement,
  React.TdHTMLAttributes<HTMLTableCellElement>
>(({ className, ...props }, ref) => (
  <td
    ref={ref}
    className={cn("p-4 align-middle [&:has([role=checkbox])]:pr-0", className)}
    {...props}
  />
));
TableCell.displayName = "TableCell";

export default TableComponent; 
