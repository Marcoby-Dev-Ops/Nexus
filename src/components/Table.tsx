import React from 'react';
import type { ReactNode } from 'react';
import PropTypes from 'prop-types';

/**
 * @interface TableColumn
 * @description Table column definition for the Table component.
 */
export interface TableColumn<T> {
  header: string;
  accessor: keyof T;
}

/**
 * @interface TableProps
 * @description Props for the Table component.
 */
export interface TableProps<T> {
  columns: TableColumn<T>[];
  data: T[];
  className?: string;
}

/**
 * @name Table
 * @description A table component for displaying tabular data.
 * @param {TableProps<T>} props - The props for the component.
 * @returns {React.ReactElement} The rendered Table component.
 */
export function Table<T extends object>({ columns, data, className = '' }: TableProps<T>): React.ReactElement {
  return (
    <table className={`min-w-full divide-y divide-border ${className}`}>
      <thead className="bg-muted">
        <tr>
          {columns.map((col) => (
            <th key={col.header} className="px-4 py-2 text-left text-xs font-medium text-muted-foreground uppercase tracking-wider">{col.header}</th>
          ))}
        </tr>
      </thead>
      <tbody className="bg-background divide-y divide-border">
        {data.map((row, idx) => (
          <tr key={idx}>
            {columns.map((col) => (
              <td key={col.header} className="px-4 py-2 whitespace-nowrap">{row[col.accessor] as ReactNode}</td>
            ))}
          </tr>
        ))}
      </tbody>
    </table>
  );
}

Table.propTypes = {
  columns: PropTypes.arrayOf(
    PropTypes.shape({
      header: PropTypes.string.isRequired,
      accessor: PropTypes.string.isRequired,
    })
  ).isRequired,
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
  className: PropTypes.string,
};

export default Table; 