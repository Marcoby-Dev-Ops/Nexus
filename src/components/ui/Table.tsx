import React from 'react';
import { cn } from '@/lib/utils';

export interface Column {
  Header: string;
  accessor: string;
}

export interface TableProps {
  columns: Column[];
  data: any[];
  className?: string;
}

/**
 * Simple table component for 1.0
 */
const Table: React.FC<TableProps> = ({ columns, data, className }) => {
  return (
    <div className={cn('overflow-x-auto', className)}>
      <table className="min-w-full divide-y divide-border">
        <thead className="bg-muted">
          <tr>
            {columns.map((column, index) => (
              <th 
                key={index}
                className="px-6 py-3 text-left text-xs font-medium text-foreground uppercase tracking-wider"
              >
                {column.Header}
              </th>
            ))}
          </tr>
        </thead>
        <tbody className="bg-background divide-y divide-border">
          {data.map((row, rowIndex) => (
            <tr key={rowIndex} className="hover:bg-muted/50">
              {columns.map((column, colIndex) => (
                <td 
                  key={colIndex}
                  className="px-6 py-4 whitespace-nowrap text-sm text-foreground"
                >
                  {row[column.accessor]}
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default Table; 