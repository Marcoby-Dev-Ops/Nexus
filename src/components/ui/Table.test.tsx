/**
 * @file Table.test.tsx
 * @description Unit and snapshot tests for the Table component.
 */
import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Table from './Table';

const columns = [
  { Header: 'Name', accessor: 'name' },
  { Header: 'Age', accessor: 'age' },
];
const data = [
  { name: 'Alice', age: 30 },
  { name: 'Bob', age: 25 },
];

describe('Table', () => {
  it('renders table headers and data', () => {
    render(<Table columns={columns} data={data} />);
    expect(screen.getByText('Name')).toBeInTheDocument();
    expect(screen.getByText('Age')).toBeInTheDocument();
    expect(screen.getByText('Alice')).toBeInTheDocument();
    expect(screen.getByText('Bob')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<Table columns={columns} data={data} />);
    expect(asFragment()).toMatchSnapshot();
  });
}); 