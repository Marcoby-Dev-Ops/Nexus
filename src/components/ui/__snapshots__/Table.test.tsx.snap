// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="overflow-x-auto"
  >
    <table
      class="min-w-full divide-y divide-border"
    >
      <thead
        class="bg-muted"
      >
        <tr>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-foreground uppercase tracking-wider"
          >
            Name
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-foreground uppercase tracking-wider"
          >
            Age
          </th>
        </tr>
      </thead>
      <tbody
        class="bg-background divide-y divide-border"
      >
        <tr
          class="hover:bg-muted/50"
        >
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-foreground"
          >
            Alice
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-foreground"
          >
            30
          </td>
        </tr>
        <tr
          class="hover:bg-muted/50"
        >
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-foreground"
          >
            Bob
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-foreground"
          >
            25
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
